spring-cloud-stream/src/main/java/org/springframework/cloud/stream/annotation/EnableBinding.java

```diff
@@ -37,13 +36,14 @@
  * @author Dave Syer
  * @author Marius Bogoevici
  * @author David Turanski
+ * @author Soby Chacko
  */
 @Target({ ElementType.TYPE, ElementType.ANNOTATION_TYPE })
 @Retention(RetentionPolicy.RUNTIME)
 @Documented
 @Inherited
 @Configuration
-@Import({ BindingServiceConfiguration.class, BindingBeansRegistrar.class, BinderFactoryConfiguration.class})
+@Import({ BindingBeansRegistrar.class, BinderFactoryConfiguration.class})
```

Comment:
Unless I miss anything specific, the `BindingServiceConfiguration` was intended to be associated with `EnableBinding` only and all the beans of `BindingServiceConfiguration` would be of no value without `EnableBinding`.

Reply:
Hi, @ilayaperumalg with the boot 2.1 changes that don't allow bean name overriding kind of started us to do it in this direction. There were `ConditionalOnMissingBean` annotations present on beans in `BindingServiceConfigurartion` and this was not an auto-configurable class. Since, it is not advisable to add `ConditionalOnMissing..` annotations on non auto-configurable classes, it was not guaranteed that it would work. In fact, we ran into that scenario and that's the reason. Downstream binders, sample apps etc. ran without issues. Seems to me that `BindingServiceConfiguration` is such a fundamental configuration class and could be benefited from auto-configuration. What do you think?

Reply:
ok, Thanks for the clarification @sobychacko. With these changes, the `BinderTypeRegistry` is always going to be `DefaultBinderTypeRegistry` since we have now removed the `@ConditionalOnMissingBean(BinderTypeRegistry.class)` for the `BinderTypeRegistry` bean. Are we ok with that?

I thought both the `BindingServiceConfiguration` and the `BinderFactoryConfiguration` go hand in hand and the `BinderFactoryConfiguration` can also be moved into `BindingServiceConfiguration` as part of the autoconfiguration.

Reply:
@ilayaperumalg ,

One more time: as long as your `@Configuration` is not included into the `spring.factories` `EnableAutoConfiguration` property, all those `@Conditional...` don't work. And that's how we got all those problems with the new `BeanDefinitionOverrideException`.

The point is that this `BinderTypeRegistry` bean is really appear in our app only when we have that `@EnableBinding`, and as early as possible. And that's how that our new `BindingServiceConfiguration` (I believe we need to rename it into an `...Auto...`) depends on the `@ConditionalOnBean(value = BinderTypeRegistry.class, search = SearchStrategy.CURRENT)`. It's not going to appear in our context if we don't have `@EnableBinding`.

I don't think that `BinderTypeRegistry` is something intended for overriding. If you would like to have some other impl, you would need a custom `@EnableBinding` as well.

The bean override didn't work before properly and I'm not sure it is going to be so critical that we don't allow to override a `BinderTypeRegistry` since now...

Reply:
The Copyright has to be update.
BTW, for all the affected classes...

Reply:
@ilayaperumalg yes, we're ok with "`BinderTypeRegistry` always being a  `DefaultBinderTypeRegistry`". Soby and I have discussed it quite a bit and we don't envision it as an extension point any longer.

Reply:
@sobychacko @artembilan @olegz Thank you for the explanation! I get the intention for this change now.

One suggestion about this:

> The point is that this BinderTypeRegistry bean is really appear in our app only when we have that @EnableBinding, and as early as possible. And that's how that our new BindingServiceConfiguration (I believe we need to rename it into an ...Auto...) depends on the @ConditionalOnBean(value = BinderTypeRegistry.class, search = SearchStrategy.CURRENT). It's not going to appear in our context if we don't have @EnableBinding.

Maybe it is worth having a different bean or a marker interface to check on initializing the `BindingServiceConfiguration` as auto-configuration instead of using `ConditionOnBean` with `BinderTypeRegistry`.  I believe the bean `BinderTypeRegistry` is just the inner bean (the holder of the binder types) for `BinderFactoryConfiguration` and can't be relied upon initializing the whole `BindingServiceConfiguration` itself.